src/components/DropChild.tsx

```diff
@@ -54,6 +56,20 @@ const DropChild = ({ depth, last, path, simplePath }: DropChildProps) => {
   const { isHovering, dropTarget } = useDragAndDropSubThought({ path, simplePath })
   useHoveringPath(path, !!isHovering, DropThoughtZone.SubthoughtsDrop)
 
+  const parentPath = rootedParentOf(
+    useSelector(state => state),
+    path,
+  )
+
+  // Get siblings of the current thought (or root thoughts if no parent)
+  const siblings = useSelector(state => getChildrenSorted(state, head(parentPath)))
+
+  // Check if this thought is the last in its parent's list of children (or in the root context)
+  const isLastChild = head(siblings).id === head(path)
+
+  // Calculate the height for the child thought over cliff
+  const dropTargetHeight = isLastChild ? calculateCliffDropTargetHeight({ depth }) : 0
```

Comment:
So this would only work if we only have visible children of a single context I believe, unlike in this case where multiple context children are expanded.

Reply:
Yes. I think checking last child of a given parent is the wrong approach. The LayoutTree should probably pass down if it is the last visible thought in the entire rendered tree.

Reply:
I think this was done but it seems commented out and implementation removed? from the `DropEnd` in `LayoutTree`.

```
<DropEnd
  depth={pathEnd.length}
  path={pathEnd}
  cliff={cliff}
  // last={!nextChildId}
/>
```

Reply:
Or I think if it can be calculated in `VirtualThought`, where we are already checking if the thought `isVisible`, we can pass the nextThought to `VirtualThought` and check if its either hidden or null and if current `isVisible` to determine the last rendered thought.

Reply:
In general we always want functions or components to be given as little information as possible for them to do their job. Doing this reduces the coupling between the function and the rest of the app. Components are on a "need-to-know" basis.

If the logic for `isLastVisible` were all derivable from the Redux state, then at least it would be encapsulable within a selector. However `nextThought` is a product of linearizing and positioning that happens within LayoutTree. `VirtualThought` should not have to know about this linearizing process. The only thing it cares about is whether it is the last visible child in the render tree, and that information is properly determined by the LayoutTree.

Reply:
I see, then it has to be then calculated within the `useMemo` render phase and returned from `treeThoughtPositioned` but of course this logic cannot be completely used there.

Reply:
is the `node.visibleChildrenKeys` not available? it seems missing from the `treeThoughtPositioned` items most of the time.

Reply:
It's only used in table view, [as documented](https://github.com/cybersemics/em/blob/57c332047a3517ff832f037544f9c6438c13f364/src/components/LayoutTree.tsx#L65-L67). I would not recommend using it for other purposes.

Reply:
I think similar logic can be used in `useMemo` map function the only problem I see is the use of `store.getState()` for `calculateAutoFocus`. If it does not do any harm, this could work fine.

Reply:
> It's only used in table view, [as documented](https://github.com/cybersemics/em/blob/57c332047a3517ff832f037544f9c6438c13f364/src/components/LayoutTree.tsx#L65-L67). I would not recommend using it for other purposes.

No I was just asking, it did not seem useful at first glance, but I realize it is not going to be helpful.

Reply:
> I think similar logic can be used in `useMemo` map function the only problem I see is the use of `store.getState()` for `calculateAutoFocus`. If it does not do any harm, this could work fine.

Calling `store.getState()` directly in components is not recommended. We generally need to rely on `useSelector` to ensure the value gets recalculated when the state changes.

`linearizeTree` has access to the state, so you could call `calculateAutofocus` there, though you have to be careful to only call it when needed as it is a heavy function.

Another possibility is to move `calculateAutofocus` up a level to `LayoutTree` instead of `VirtualThought`. It can be calculated in the linearize step. Then you would be able to generate `isLastVisible` there.

Reply:
I was saying the same but in the `useMemo` function not in the `VirutualThought` btw. I did not do it `linearizeTree` but in `useMemo` and it is working fine, but I see that `linearizeTree` has access to `state` so it seems more logical to put it there instead of using `store.getState` in `useMemo`.

Reply:
Yes, exactly